### ISSUE TITLE
Questions/Discussion

### ISSUE THREAD

**Issue by 8carlosf** [Babylonpartners/fastText_multilingual#7]
First question is, how can I align FastText vectors with GloVe, this should be easy, given that words in one side should have a direct match on the other.
If I do that, shouldn't it also be easy to combine the existing matrix's with the new FastText to GloVe matrix, making it only one operation to align any language with Glove? (basically creating alignments for every language to GloVe instead of FastText, just to be able to test stuff in pre-trained GloVe models)
How can we do that?

Second, wouldn't it be interesting to do multiple (grouped) alignments for each Language Family or Sub-Family. Why? If we train a English to Spanish seq2seq translator, and Spanish is aligned with a similar language (like Portuguese and other Latin Languages only), we could get better results when replacing the Spanish vectors with the Portuguese ones.

Please let me know if my ideas are not well explained, I think that this is a interesting discussion, and could simplify the process for lot's of tasks (train in one language and use in another, with really good results).

Btw, incredible work :) thank you for sharing

**Comment by sls56** [comment]
Hi Carlos,
Yes, it would be easy to align a set of FastText vectors with a set of Glove vectors from the same language. We provided example code for aligning different representations in "align_your_own.ipynb". In that example we constructed our training dictionary from the identical character strings occurring in two different languages (Amazingly this works!). But this is also exactly what you want to do when aligning a language to itself, so just feed in the two different representations of the same language and everything should work fine.

You are right that once you have, for example, aligned pre-trained English Glove vectors with the English FastText vectors, you can then immediately infer the transformation to any other language we provide by simply multiplying the two orthogonal transformations. Of course this will introduce some noise, but it should still work acceptably well.

And  yes, you can see in the performance matrix that similar languages cluster; so it would be natural to define groups of similar languages and obtain multilingual word vectors for each group!

Please fire any more questions at me, delighted you enjoyed the work!

Best wishes,
Sam

**Comment by 8carlosf** [comment]
I have another question. Can we load .bin models and align them? so that when we use fasttext for Out-of-Vocabulary words it will automatically spit out aligned vectors..

About the selected words that we use to compute the alignment, if we could get native speakers to help (should be easy, because we don't need that many words to compute a good alignment) we could make a more careful selection, this being, only use words that when translated from A to B also produce a single word (we may lose quality doing averages) and also, if we had native speakers they could judge if the translation has almost always the same meaning, or if it is one of that words that depends a lot on context (etc). (not sure if I'm just complicating, or if this could improve stuff, but I can try it with Portuguese when I get the time).

Also, may be interesting to think about other types of matrix transformations that will better align the vectors but losing monolingual relationships (I'm not sure if there are such matrix operations, but maybe you know some tricks), and then, use that as a pre-train to re-compute the fasttext vectores. That way the vectors should be already in the correct neighbourhood, the train would be fast, and the alignment may be even better.

Again, I think the potential in your work is in pre-training encoding/embeddings, so you train a English sentence encoder for a given task, you can just use it as seed in another language, hot swapping the vectors and re-training it in that language (this is really useful, because many language have insufficient data to do a good full train).

**Comment by sls56** [comment]
I don't think we can currently load and rotate the .bin models directly; but you can extract word vectors from the .bin; and then apply the transformation.

So certainly improved dictionaries would improve the transformations a bit; but I think there are more gains to be had (and more interesting research to be done :) ) by improving the methods.

Yes: so this work is mostly focused on the problem of aligning different pre-trained embeddings when there is minimal information about the overlap between the two. But another interesting research problem is how to learn an embedding for a target language from limited data, using a well-trained embedding of a source language and abundant information about translation pairs. In principle this would be fairly simple to implement (e.g. one could add a regulariser to word2vec which minimises the squared distance between a target word and its known source translation vector-> thus using the source language as a scaffold).

**Comment by TheGlobalist** [comment]
Hi @sls56, I'd like to use this issue (instead opening a new one) for asking you a question as well.
What's the use for the alignment_matrix folder? Could I just parse them, load in a numpy array and concatenate them with n languages?
